libsignal-service/src/sender.rs

```diff
@@ -331,51 +331,57 @@ where
         }
 
         // try to send the original message to all the recipient's devices
-        let mut results = vec![
-            self.try_send_message(
+        let result = self
+            .try_send_message(
                 *recipient,
                 unidentified_access.as_ref(),
                 &content_body,
                 timestamp,
                 include_pni_signature,
                 online,
             )
-            .await,
-        ];
-
-        match (&content_body, &results[0]) {
-            // if we sent a data message and we have linked devices, we need to send a sync message
-            (
-                ContentBody::DataMessage(message),
-                Ok(SentMessage { needs_sync, .. }),
-            ) if *needs_sync || self.is_multi_device().await => {
-                tracing::debug!("sending multi-device sync message");
-                let sync_message = self
-                    .create_multi_device_sent_transcript_content(
-                        Some(recipient),
-                        Some(message.clone()),
-                        timestamp,
-                        &results,
-                    );
-                self.try_send_message(
-                    self.local_aci,
-                    None,
-                    &sync_message,
+            .await;
+
+        let needs_sync = match &result {
+            Ok(SentMessage { needs_sync, .. }) => *needs_sync,
+            _ => false,
+        };
+
+        if needs_sync || self.is_multi_device().await {
+            let data_message = match &content_body {
+                ContentBody::DataMessage(m) => Some(m.clone()),
+                _ => None,
+            };
+            let edit_message = match &content_body {
+                ContentBody::EditMessage(m) => Some(m.clone()),
+                _ => None,
+            };
+            tracing::debug!("sending multi-device sync message");
+            let sync_message = self
+                .create_multi_device_sent_transcript_content(
+                    Some(recipient),
+                    data_message,
+                    edit_message,
                     timestamp,
-                    false,
-                    false,
-                )
-                .await?;
-            },
-            _ => (),
+                    Some(&result),
+                );
+            self.try_send_message(
+                self.local_aci,
+                None,
+                &sync_message,
+                timestamp,
+                false,
+                false,
+            )
+            .await?;
         }
 
         if end_session {
             let n = self.protocol_store.delete_all_sessions(recipient).await?;
             tracing::debug!("ended {} sessions with {}", n, recipient.uuid);
         }
 
-        results.remove(0)
+        result
```

Comment:
Maybe we should add the sync result here too, like it is done in send to group?

Reply:
I think that's the reason `results` used to be a Vec, but at this point I'm not sure anymore.

Reply:
This would break the result type of the public API. If that's ok, I can change it to returning a vec.

Reply:
I'd like for @direc85's opinion too, since he last touched these things iirc, but I think it makes sense. Maybe we should cross-check with what SA/SD do.

Reply:
Every time we touch this part of the code we improve something and break another 😂 this is just bound to happen at this point. Hope your fix is the final one.

Reply:
I'm in favor of returning a `Vec` here. What to do with the results is another question... Is it libsignals problem? Or the clients'?

Reply:
I checked the impl in Signal Android. They only return the results of the actual message sending not the result of the sync message. In particular, when sending a message to a single recipient there is a single result, when sending to a group there is a list of results.

I think we should keep it aligned. I will remove the sync result from the list of result when sending to a group.